malsim/mal_simulator.py

```diff
@@ -556,57 +600,54 @@ def step(
+        step_all_compromised_nodes: set[AttackGraphNode] = set()
+        step_agent_compromised_nodes: dict[str, set] = {}
+        step_enabled_defenses: set[AttackGraphNode] = set()
+        step_nodes_made_unviable: set[AttackGraphNode] = set()
 
         # Perform defender actions first
         for defender_state in self._get_defender_agents():
             enabled, unviable = self._defender_step(
                 defender_state, actions.get(defender_state.name, [])
             )
             step_enabled_defenses |= enabled
-            step_unviable_nodes |= unviable
+            step_nodes_made_unviable |= unviable
 
         # Perform attacker actions afterwards
         for attacker_state in self._get_attacker_agents():
-            compromised = self._attacker_step(
+            agent_compromised = self._attacker_step(
                 attacker_state, actions.get(attacker_state.name, [])
             )
-            step_compromised_nodes |= compromised
-
-            # Calculate attacker reward and termination
-            attacker_state.reward = self._attacker_reward(attacker_state)
-            attacker_state.truncated = self.cur_iter >= self.max_iter
-            attacker_state.terminated = (
-                self._attacker_is_terminated(attacker_state)
-            )
-
-        # Calculate defender rewards and terminations
-        # (depends on attackers, so must be done after all steps)
-        for defender_state in self._get_defender_agents():
-            defender_state.step_all_compromised_nodes = step_compromised_nodes
-            defender_state.reward = self._defender_reward(defender_state)
-            defender_state.truncated = self.cur_iter >= self.max_iter
-            defender_state.terminated = self._defender_is_terminated(
-                self._get_attacker_agents()
-            )
+            step_all_compromised_nodes |= agent_compromised
+            step_agent_compromised_nodes[attacker_state.name] = agent_compromised
 
-        # Remove agents that are terminated or truncated
+        # Calculate new agent states
         for agent_name in self._alive_agents.copy():
             agent_state = self._agent_states[agent_name]
-            agent_state.truncated = self.cur_iter >= self.max_iter
+
+            if isinstance(agent_state, MalSimAttackerState):
+                self._update_attacker_state(
+                    agent_state,
+                    step_agent_compromised_nodes.get(agent_name, set()),
+                    step_nodes_made_unviable
+                )
+
+            elif isinstance(agent_state, MalSimDefenderState):
+                self._update_defender_state(
+                    agent_state,
+                    step_all_compromised_nodes,
+                    step_enabled_defenses,
+                    step_nodes_made_unviable
+                )
+            else:
+                raise TypeError(
+                    f"Unknown agent class {agent_state.__class__}"
+                )
```

Comment:
I don't think this branch is needed. The only ones that handle the self._agent_states variable is us in this class. It is private and users of this class should not touch it. If they do, their problem.